File: cpp/ecdsa-op-count.py
# 
# This script calculates and prints the approximate number of 32-bit arithmetic operations
# needed to perform elliptic curve point multiplication, in the C++ implementation.
# For Python 2 and 3. Run with no arguments.
# 
# Bitcoin cryptography library
# Copyright (c) Project Nayuki
# 
# https://www.nayuki.io/page/bitcoin-cryptography-library
# https://github.com/nayuki/Bitcoin-Cryptography-Library
# 


# ---- Calculate operation counts ----

# Function call overhead
funcOvh = 1

# Uint256 methods
uiCopy        = funcOvh + 8*2
uiReplace     = funcOvh + 8*4
uiSwap        = funcOvh + 8*7
uiEquals      = funcOvh + 3 + 8*2
uiLessThan    = funcOvh + 2 + 8*6
uiAdd         = funcOvh + 3 + 8*9
uiSubtract    = funcOvh + 3 + 8*9
uiShiftLeft1  = funcOvh + 2 + 8*4
uiShiftRight1 = funcOvh + 6 + 8*8
uiReciprocal  = funcOvh + 4*uiCopy + uiReplace + 512*(11 + 2*uiSwap + 2*uiAdd + 2*uiSubtract + 2*uiShiftRight1)

# FieldInt methods
fiCopy       = funcOvh + uiCopy
fiReplace    = funcOvh + uiReplace
fiEquals     = funcOvh + uiEquals
fiLessThan   = funcOvh + uiLessThan
fiAdd        = funcOvh + 3 + fiLessThan + uiAdd + uiSubtract
fiSubtract   = funcOvh + 1 + uiAdd + uiSubtract
fiMultiply2  = funcOvh + 3 + uiSubtract + uiShiftLeft1
fiMultiply   = funcOvh + 754 + (1 + 24*28) + (1 + 16*26) + (1 + 9*8) + 10 + 3 + uiLessThan + uiSubtract
fiSquare     = funcOvh + fiMultiply
fiReciprocal = funcOvh + uiReciprocal

# CurvePoint methods
cpCopy      = funcOvh + 3*fiCopy
cpReplace   = funcOvh + 3*fiReplace
cpIsZero    = funcOvh + 2 + 3*fiEquals
cpEquals    = funcOvh + 1 + 3*fiEquals
cpTwice     = funcOvh + 1 + 10*fiCopy + fiEquals + fiAdd + 3*fiSubtract + 5*fiMultiply2 + 7*fiMultiply + 4*fiSquare + cpReplace + cpIsZero
cpAdd       = funcOvh + 10 + 10*fiCopy + 3*fiReplace + 2*fiEquals + fiAdd + 5*fiSubtract + 13*fiMultiply + 2*fiSquare + cpCopy + 3*cpReplace + 2*cpIsZero + cpTwice
cpMultiply  = funcOvh + 18*cpCopy + cpTwice + 13*(cpCopy + cpAdd) + 64*(36 + cpCopy + 16*cpReplace + 4*cpTwice + cpAdd) - 4*cpTwice
cpNormalize = funcOvh + 1 + fiCopy + 2*fiReplace + 3*fiEquals + 2*fiMultiply + fiReciprocal + cpCopy + cpReplace
cpIsOnCurve = funcOvh + 2 + 2*fiCopy + fiEquals + 2*fiAdd + fiMultiply + 2*fiSquare + cpIsZero

# Ecdsa methods
edMulModOrder = funcOvh + 1 + 2*uiCopy + 256*(8 + 2*uiLessThan + uiAdd + 2*uiSubtract + uiShiftLeft1)
edSign        = funcOvh + 4 + 7*uiCopy + uiReplace + 3*uiEquals + 4*uiLessThan + uiAdd + 3*uiSubtract + uiReciprocal + cpCopy + cpMultiply + cpNormalize + 2*edMulModOrder
edVerify      = funcOvh + 11 + 5*uiCopy + uiEquals + 5*uiLessThan + uiSubtract + uiReciprocal + fiEquals + 3*cpCopy + 2*cpIsZero + cpAdd + 3*cpMultiply + cpNormalize + cpIsOnCurve + 2*edMulModOrder



# ---- Print numbers ----

vargroups = [
	["uiCopy", "uiReplace", "uiSwap", "uiEquals", "uiLessThan", "uiAdd", "uiSubtract", "uiShiftLeft1", "uiShiftRight1", "uiReciprocal"],
	["fiCopy", "fiReplace", "fiEquals", "fiLessThan", "fiAdd", "fiSubtract", "fiMultiply2", "fiMultiply", "fiSquare", "fiReciprocal"],
	["cpCopy", "cpReplace", "cpIsZero", "cpEquals", "cpTwice", "cpAdd", "cpMultiply", "cpNormalize", "cpIsOnCurve"],
	["edMulModOrder", "edSign", "edVerify"],
]

def inttostr_with_groups(x):
	result = str(x)
	for i in range(len(result) - 3, 0, -3):
		result = result[ : i] + " " + result[i : ]
	return result

for group in vargroups:
	for varname in group:
		print("{:>11}  {}".format(inttostr_with_groups(globals()[varname]), varname))
	print("")

File: cpp/EcdsaOpCount.cpp
/* 
 * A runnable main program that calculates and prints the approximate
 * number of 32-bit arithmetic operations needed to perform
 * elliptic curve point multiplication, in this C++ implementation.
 * 
 * Bitcoin cryptography library
 * Copyright (c) Project Nayuki
 * 
 * https://www.nayuki.io/page/bitcoin-cryptography-library
 * https://github.com/nayuki/Bitcoin-Cryptography-Library
 */

#include <cstddef>
#include <cstdlib>
#include <iostream>
#include <string>
#include "CountOps.hpp"
#include "CurvePoint.hpp"
#include "Ecdsa.hpp"
#include "FieldInt.hpp"
#include "Sha256.hpp"
#include "Sha256Hash.hpp"
#include "Uint256.hpp"


static long long opsCount;


void countOps(long n) {
	opsCount += n;
}


static void printOps(const char *name);
static void doUint256();
static void doFieldInt();
static void doCurvePoint();
static void doEcdsa();


int main() {
	doUint256();
	doFieldInt();
	doCurvePoint();
	doEcdsa();
	return EXIT_SUCCESS;
}


static void doUint256() {
	{
		Uint256 x = Uint256::ONE;
		Uint256 y = Uint256::ONE;
		opsCount = 0;
		x.replace(y, 1);
		printOps("uiReplace");
	}
	{
		Uint256 x = Uint256::ONE;
		Uint256 y = Uint256::ONE;
		opsCount = 0;
		x.swap(y, 1);
		printOps("uiSwap");
	}
	{
		Uint256 x = Uint256::ONE;
		Uint256 y = Uint256::ONE;
		opsCount = 0;
		x == y;
		printOps("uiEquals");
	}
	{
		Uint256 x = Uint256::ONE;
		Uint256 y = Uint256::ONE;
		opsCount = 0;
		x < y;
		printOps("uiLessThan");
	}
	{
		Uint256 x = Uint256::ONE;
		Uint256 y = Uint256::ONE;
		opsCount = 0;
		x.add(y);
		printOps("uiAdd");
	}
	{
		Uint256 x = Uint256::ONE;
		Uint256 y = Uint256::ONE;
		opsCount = 0;
		x.subtract(y);
		printOps("uiSubtract");
	}
	{
		Uint256 x = Uint256::ONE;
		opsCount = 0;
		x.shiftLeft1();
		printOps("uiShiftLeft1");
	}
	{
		Uint256 x = Uint256::ONE;
		opsCount = 0;
		x.shiftRight1();
		printOps("uiShiftRight1");
	}
	{
		Uint256 x = Uint256::ONE;
		Uint256 y = CurvePoint::ORDER;
		opsCount = 0;
		x.reciprocal(y);
		printOps("uiReciprocal");
	}
	std::cout << std::endl;
}


static void doFieldInt() {
	{
		FieldInt x(Uint256::ONE);
		FieldInt y(Uint256::ONE);
		opsCount = 0;
		x.replace(y, 1);
		printOps("fiReplace");
	}
	{
		FieldInt x(Uint256::ONE);
		FieldInt y(Uint256::ONE);
		opsCount = 0;
		x == y;
		printOps("fiEquals");
	}
	{
		FieldInt x(Uint256::ONE);
		FieldInt y(Uint256::ONE);
		opsCount = 0;
		x < y;
		printOps("fiLessThan");
	}
	{
		FieldInt x(Uint256::ONE);
		FieldInt y(Uint256::ONE);
		opsCount = 0;
		x.add(y);
		printOps("fiAdd");
	}
	{
		FieldInt x(Uint256::ONE);
		FieldInt y(Uint256::ONE);
		opsCount = 0;
		x.subtract(y);
		printOps("fiSubtract");
	}
	{
		FieldInt x(Uint256::ONE);
		opsCount = 0;
		x.multiply2();
		printOps("fiMultiply2");
	}
	{
		FieldInt x(Uint256::ONE);
		FieldInt y(Uint256::ONE);
		opsCount = 0;
		x.multiply(y);
		printOps("fiMultiply");
	}
	{
		FieldInt x(Uint256::ONE);
		opsCount = 0;
		x.square();
		printOps("fiSquare");
	}
	{
		FieldInt x(Uint256::ONE);
		opsCount = 0;
		x.reciprocal();
		printOps("fiReciprocal");
	}
	std::cout << std::endl;
}


static void doCurvePoint() {
	{
		CurvePoint x = CurvePoint::G;
		CurvePoint y = CurvePoint::G;
		opsCount = 0;
		x.replace(y, 1);
		printOps("cpReplace");
	}
	{
		CurvePoint x = CurvePoint::G;
		opsCount = 0;
		x.isZero();
		printOps("cpIsZero");
	}
	{
		CurvePoint x = CurvePoint::G;
		CurvePoint y = CurvePoint::G;
		opsCount = 0;
		x == y;
		printOps("cpEquals");
	}
	{
		CurvePoint x = CurvePoint::G;
		opsCount = 0;
		x.twice();
		printOps("cpTwice");
	}
	{
		CurvePoint x = CurvePoint::G;
		CurvePoint y = CurvePoint::G;
		opsCount = 0;
		x.add(y);
		printOps("cpAdd");
	}
	{
		CurvePoint x = CurvePoint::G;
		Uint256 y = Uint256::ONE;
		opsCount = 0;
		x.multiply(y);
		printOps("cpMultiply");
	}
	{
		CurvePoint x = CurvePoint::G;
		opsCount = 0;
		x.normalize();
		printOps("cpNormalize");
	}
	{
		CurvePoint x = CurvePoint::G;
		opsCount = 0;
		x.isOnCurve();
		printOps("cpIsOnCurve");
	}
	std::cout << std::endl;
}


static void doEcdsa() {
	{
		Uint256 privKey = Uint256::ONE;
		Sha256Hash msgHash = Sha256::getHash(nullptr, 0);
		Uint256 nonce = Uint256::ONE;
		Uint256 outR, outS;
		opsCount = 0;
		Ecdsa::sign(privKey, msgHash, nonce, outR, outS);
		printOps("edSign");
	}
	{
		CurvePoint pubKey = CurvePoint::G;
		Sha256Hash msgHash = Sha256::getHash(nullptr, 0);
		Uint256 r = Uint256::ONE;
		Uint256 s = Uint256::ONE;
		opsCount = 0;
		Ecdsa::verify(pubKey, msgHash, r, s);
		printOps("edVerify");
	}
	std::cout << std::endl;
}


static void printOps(const char *name) {
	std::string s = std::to_string(opsCount);
	while (s.size() < 9)
		s.insert(0, " ", 1);
	for (std::size_t i = s.size(); i >= 4; i -= 3)
		s.insert(i - 3, " ", 1);
	std::cout << s << "  " << name << std::endl;
}
